Add Travis CI support

ulugbekna opened the pull request:
Added Travis CI support based on [ocaml-ci-scripts repo](https://github.com/ocaml/ocaml-ci-scripts). Solves #5 

1) Even with push rights, it seems like I cannot add Travis CI support to this repo: I have neither Settings tab on the repo nor this repo in my travis account on the [travis website](https://travis-ci.org). 

2) I tested the build on my own repo, the _setup part goes well_, but `make tests` fails with the error below: 
```
# === Error ======================================================================
# 
# Test generated term passes OCaml's typecheck errored on (3 shrink steps):
# 
# Some (0)
# 
# exception Sys_error("testdir/ocamltest.ml: No such file or directory")
# 
# ================================================================================
# failure (0 tests failed, 1 tests errored, ran 2 tests)
# make: *** [tests] Error 1
```
Running the same command on my machine does not produce this error. For some reason, `ocamltest.ml` file is missing.

3) Also to note, the build times are quite large:
 ~15 min for linux
 ~20 min for macos

gasche commented:
I haven't looked in details yet, but two remarks:

- There is a mix of commits in this PR that are part of #7. Unless there is a strong dependency between the two PRs (the latter cannot work without the former; which I think is not the case here), it's better to send separate/independent PRs than stack one on top of each other, because it makes reviewing easier (here I see the diffs for both PRs combined). This requires maintaining several independent branches on top of `master`. From the current state, you can probably do this by running, from the `add_ci` branch: `git rebase --onto master additional_refactorings add_ci`, which means: "rebase the work between additional_refactoring (excluded) and add_ci (included) on top of master".

- I think that the long build times come from the use of `.travis-docker`, which builds a Docker image from scratch on each CI run. `travis-ocaml` and `travis-opam` give faster CI builds, because they reuse the binary `opam` package available on the OS, and only build the OCaml part each time. I would recommend looking at how other projects do their CI, see for example `ppx_deriving/.travis.yml`, which is super-simple: it downloads and runs the `travis-opam` switch, which builds and runs the test of the local `.opam` package (in our case `efftester.opam`, which should get a `{with-test}` rule, see [ppx_deriving/ppx_deriving.opam](https://github.com/ocaml-ppx/ppx_deriving/blob/master/ppx_deriving.opam) (it could be just `[make "test"] {with-test}`).

ulugbekna commented:
- Removed unrelated commits
- Added travis ci support (build takes ~7 min; build is successful)
- Renamed `testdir` to `generated_tests` because we might eventually want to move our source code test(s) outside `src` in, say, `tests` folder in the root. 

Note: Still need the repo's owner's help to finalize travis ci setup.

gasche commented:
So I did click on some buttons on my side and it looks like Travis CI is enabled *twice*, in two different ways, for this repository -- or at least for this pull request. I will click on more buttons now to try to reduce the duplication of effort.

gasche commented:
On the side of the actual PR, I think that the changes are good: the CI code is now very simple, and the `tests` support in dune is great. I will merge once the CI is green.

(Sometimes I wish we were using Gitlab instead of Github, as I'm trying to lessen my dependency on Github as a unique hosting platform. But following what the authors of the tool choose is also a sensible approach, and getting used to this dominant platform is also a good learning experience.)

gasche commented:
Merged, thanks!
